### PR TITLE
Reach 100% code coverage

### PR DESCRIPTION
**Pull request by jamadden**
Mostly this means new tests. Sometimes its means "no cover" markers. In a few cases it means using a shorter, more declarative way to write the code (e.g., `@total_ordering` or `ABCMeta`) or unifying python2/python 3 branches as much as possible. Sometimes we got lucky and could drop code meant for 2.6 or earlier.

There are a few genuine bugfixes in here that were revealed by getting the coverage up.

I tried to limit the use of 'no cover' as much as possible. There are a few deeply nested functions that made it difficult to avoid, though.

This replaces #12.

### CODE REVIEW

**Comment by tseaver** [comment]
@jamadden Now that it is at 100%, can you make coveralls not spew into the conversation itself?  I hate getting emails from bots. :)

**Comment by jamadden** [comment]
I sure can and did. It's on by default, and the badge sure is pretty to look at, but it probably doesn't add much value. And there was that recent bug where it was commenting for every Travis build in the matrix which was *really* annoying; fortunately that seems to have been fixed.

**Comment by mgedmin** [comment]
Making the tests fail when coverage drops below 100% (as @tseaver likes to do, which I greatly admire) is a better way to maintain test coverage than those Coveralls comments that annoy everyone and that nobody ever reads.

**Comment by jamadden** [comment]
PRs are set to fail in that case.

File: ZConfig/info.py [hunk 1]
```diff
@@ -228,6 +220,8 @@ def __init__(self, name, sectiontype, minOccurs, maxOccurs, handler,
         # handler     - handler name called when value(s) must take effect,
         #               or None
         # attribute   - name of the attribute on the SectionValue object
+        assert maxOccurs is not None
+        # because we compare it to 1 which is a Py3 error
```
Comment (mgedmin) [inline ZConfig/info.py]: BTW `getdefault` also unconditionally compares `self.maxOccurs > 1`.

File: ZConfig/schema.py [hunk 2]
```diff
@@ -97,19 +96,24 @@ def startElement(self, name, attrs):
...
+            if self._schema is None: # pragma: no cover
                 self.error(name + " element outside of schema")
             getattr(self, "start_" + name)(attrs)
         elif name in self._cdata_tags:
-            if self._schema is None:
+            if self._schema is None: # pragma: no cover
                 self.error(name + " element outside of schema")
-            if self._cdata is not None:
+            if self._cdata is not None: # pragma: no cover
+                # this should be handled by the earliel nesting check
```
Comment (mgedmin) [inline ZConfig/schema.py]: Typo: earliel -> earlier

File: ZConfig/schema.py [hunk 3]
```diff
@@ -234,7 +239,8 @@ def get_key_info(self, attrs, element):
         any, name, attribute = self.get_name_info(attrs, element)
         if any == '*':
             self.error(element + " may not specify '*' for name")
-        if not name and any != '+':
+        if not name and any != '+': # pragma: no cover
+            # Con we even get here?
```
Comment (mgedmin) [inline ZConfig/schema.py]: Typo: Con -> Can.

File: ZConfig/tests/support.py [hunk 4]
```diff
@@ -21,21 +21,12 @@
...
-    from urllib.request import pathname2url
+from ZConfig._compat import NStringIO as StringIO
+from ZConfig._compat import pathname2url
 
 try:
     __file__
-except NameError:
+except NameError: # pragma: no cover
     import sys
     __file__ = sys.argv[0]
```
Comment (mgedmin) [inline ZConfig/tests/support.py]: I vaguely recall this sort of caution from Python 2.2/2.3 timeframe, but is it still necessary?  Under what circumstances can `__file__` be undefined in this day and age?

File: ZConfig/tests/test_config.py [hunk 5]
```diff
@@ -131,22 +138,101 @@ def test_define(self):
...
+
+        self.assertRaisesRegexp(ZConfig.ConfigurationSyntaxError,
+                                'malformed section end',
+                                self.loadtext, '<section>\n</section')
+
+        self.assertRaisesRegexp(ZConfig.ConfigurationSyntaxError,
+                                'malformed section start',
+                                self.loadtext, '<section')
+
+        # ConfigLoader.endSection raises this and its recaught and changed to a
```
Comment (mgedmin) [inline ZConfig/tests/test_config.py]: Typo: its -> it's

File: ZConfig/tests/test_datatypes.py [hunk 6]
```diff
@@ -93,14 +93,13 @@ def test_datatype_float(self):
 
         # These are not portable representations; make sure they are
         # disallowed everywhere for consistency.
```
Comment (mgedmin) [inline ZConfig/tests/test_datatypes.py]: As I mentioned above, this comment is no longer true since Python 2.6.

File: ZConfig/tests/test_datatypes.py [hunk 7]
```diff
@@ -277,10 +276,16 @@ def check(value, family, address, self=self, convert=convert):
...
-        else:
+        else: # pragma: no cover
             self.assertTrue(a1.family is None)
             self.assertTrue(a2.family is None)
 
+        convert = self.types.get('socket-binding-address')
+        check(":80",                 AF_INET, (defhost, 80))
+
+        convert = self.types.get('socket-connection-address')
+        check(":80",                 AF_INET, ("127.0.0.1", 80))
```
Comment (mgedmin) [inline ZConfig/tests/test_datatypes.py]: That is a strange amount of blank spaces between function arguments.  PEP-8 frowns on this.

File: ZConfig/tests/test_info.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+def test_suite():
+    suite = unittest.makeSuite(UnboundTestCase)
+    suite.addTest(unittest.makeSuite(BaseInfoTestCase))
+    suite.addTest(unittest.makeSuite(BaseKeyInfoTestCase))
+    suite.addTest(unittest.makeSuite(KeyInfoTestCase))
+    suite.addTest(unittest.makeSuite(SectionInfoTestCase))
+    suite.addTest(unittest.makeSuite(AbstractTypeTestCase))
+    suite.addTest(unittest.makeSuite(SectionTypeTestCase))
+    suite.addTest(unittest.makeSuite(SchemaTypeTestCase))
+    return suite
```
Comment (mgedmin) [inline ZConfig/tests/test_info.py]: Strongly suggest

```python
def test_suite()
    return unittest.defaultTestLoader.loadTestsFromName(__name__)
```

unless you have abstract classes inheriting from `unittest.TestCase` that would preclude this?

File: ZConfig/tests/test_schemaless.py [hunk 9]
```diff
@@ -18,7 +18,17 @@
...
+
+    def test_init_with_data(self):
+        s = Section(data={'k': 'v'})
+        self.assertDictEqual(s, {'k': 'v'})
 
 def test_suite():
-    return doctest.DocFileSuite("schemaless.txt", package="ZConfig")
+    suite = unittest.makeSuite(TestSection)
+    suite.addTest(doctest.DocFileSuite("schemaless.txt", package="ZConfig"))
+    return suite
```
Comment (mgedmin) [inline ZConfig/tests/test_schemaless.py]: If you switch to using `defaultTestLoader` everywhere, a nice way to mix it with doctests is
```python
def test_suite():
    return unittest.TestSuite([
        unittest.defaultTestLoader.loadTestsFromName(__name__),
        doctest.DocFileSuite(...),
    ])
```

File: ZConfig/cmdline.py [hunk 10]
```diff
@@ -155,7 +157,7 @@ def get_section_info(self, type, name):
             bk = self.basic_key(s, pos)
             if name and self._normalize_case(s) == name:
                 L.append((optpath[1:], val, pos))
-            elif bk == type:
+            elif bk == type: # pragma: no cover
```
Comment (tseaver) [inline ZConfig/cmdline.py]: Ugh, the shadowing of the `type` built-in is really ugly:  I spent more than a bit of time trying to work out how the code worked.  I wonder if we could rename the argument `type_`?

File: ZConfig/components/logger/loghandler.py [hunk 11]
```diff
@@ -25,6 +25,13 @@
 from logging.handlers import HTTPHandler, SMTPHandler
 from logging.handlers import NTEventLogHandler as Win32EventLogHandler
 
+# Export these, they're used in handlers.py
+SysLogHandler = SysLogHandler
+HTTPHandler = HTTPHandler
+SMTPHandler = SMTPHandler
+Win32EventLogHandler = Win32EventLogHandler
```
Comment (tseaver) [inline ZConfig/components/logger/loghandler.py]: I don't understand what "export these" means, or what that has to do with re-binding the name.

On another note:  couldn't `handlers.py` just import them from `logging.handlers`?

File: ZConfig/components/logger/tests/test_logger.py [hunk 12]
```diff
@@ -27,11 +27,8 @@
 from ZConfig.components.logger import handlers
 from ZConfig.components.logger import loghandler
 
-try:
-    import StringIO as StringIO
-except ImportError:
-    # Python 3 support.
-    import io as StringIO
+from ZConfig._compat import NStringIO as StringIO
```
Comment (tseaver) [inline ZConfig/components/logger/tests/test_logger.py]: Why not leave the name `NStringIO`, since we have to munge all the callers anyway?

File: ZConfig/components/logger/loghandler.py (hunk 11 above)
Comment (jamadden) [inline ZConfig/components/logger/loghandler.py]: This applies to this comment and two of the previous ones:

linters like flake8 and pylint (I run/ran both during this process) complain (rightfully) about unused imports. The way to fix that without adding `#noqa` or what have you is to *use* the import, and the simplest way to do that is to rebind it to the same name. This declarative statement seems reasonably clear to me that this name is a public part of the module definition, as opposed to a side-effect of the import.

In this particular case, it seemed important to leave the imports as they are for backwards compatibility.

File: ZConfig/cmdline.py (hunk 10 above)
Comment (jamadden) [inline ZConfig/cmdline.py]: There would be a *lot* or renames to do throughout the codebase. `type` (among others) is commonly shadowed; I think this code predates new-style classes and `type` as a builtin.

I'd be happy to see that done, but I believe it belongs in another PR.

File: ZConfig/components/logger/tests/test_logger.py (hunk 12 above)
Comment (jamadden) [inline ZConfig/components/logger/tests/test_logger.py]: It seemed like a slightly smaller change, but I admit it wasn't fully thought through. There was some evolution here. It can still change if desired.

File: ZConfig/tests/test_info.py (hunk 8 above)
Comment (jamadden) [inline ZConfig/tests/test_info.py]: I haven't seen that idiom used anywhere in this codebase (or in any of the other zopefoundation projects I've touched so far). In fact, I wasn't actually aware of it.

I used this style because that's what was used in the rest of the code base. I'd be happy to see it change to a simpler style, but I submit that would belong in a separate PR that does the same for all the test modules.

File: ZConfig/tests/test_datatypes.py (hunk 7 above)
Comment (jamadden) [inline ZConfig/tests/test_datatypes.py]: Indeed. But that matches the way this is called in the rest of the function.

File: ZConfig/tests/support.py (hunk 4 above)
Comment (jamadden) [inline ZConfig/tests/support.py]: I'm not sure. I know there is weirdness with `__file__` when the various forms of zipped packages are used, but I don't know exactly what that is for all of them (and there are a bunch now). 

If we can assume the tests are only ever run from an sdist/source checkout, then we could drop this throughout. I just wasn't sure.

File: ZConfig/cfgparser.py [hunk 13]
```diff
@@ -175,7 +170,8 @@ def replace(self, text):
             raise
 
     def error(self, message):
-        raise ZConfig.ConfigurationSyntaxError(message, self.url, self.lineno)
+        v = ZConfig.ConfigurationSyntaxError(message, self.url, self.lineno)
+        reraise(type(v), v, sys.exc_info()[2])
```
Comment (jamadden) [inline ZConfig/cfgparser.py]: Yes, this is to keep the original traceback. There are *lots* of places in the code that do `try:...except: self.error()`. Debugging those was a nightmare without keeping the original traceback. 

The simplest textual change was to change the (relatively few) `error` functions instead of changing all the callers.

File: .coveragerc [hunk 14]
```diff
@@ -0,0 +1,17 @@
+[run]
+source = ZConfig
+
+[report]
+# Coverage is run on Py2 and Py3
+# on POSIX platforms, so omit Windows-specific
+# things. Omit the main methods of test files.
+exclude_lines =
+    pragma: no cover
+    pragma no cover
```
Comment (jamadden) [inline .coveragerc]: I guess there are no occurences of 'pragma no cover' anymore so that can go away.

File: ZConfig/_compat.py [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+            if locs is None:
+                locs = frame.f_locals
+            del frame
+        elif locs is None:
+            locs = globs
+        exec("""exec code in globs, locs""")
+
+    exec_("""def reraise(tp, value, tb=None):
+    raise tp, value, tb
+""")
```
Comment (jamadden) [inline ZConfig/_compat.py]: For better or worse, that code is copied verbatim from `six` and various other zopefoundation packages that have done the same.

File: ZConfig/datatypes.py [hunk 16]
```diff
@@ -277,13 +280,10 @@ def _parse_address(self, s):
...
-        is_str = isinstance(v, basestring)
-    except NameError:
-        # Python 3 support.
-        is_str = isinstance(v, str)
+    # float() will accept both bytes and unicode on both 2 and 3
+    is_str = isinstance(v, text_type) or isinstance(v, binary_type)
     if is_str:
-        if v.lower() in ["inf", "-inf", "nan"]:
+        if v.lower() in (u"inf", u"-inf", u"nan", b'inf', b'-inf', b'nan'):
             raise ValueError(repr(v) + " is not a portable float representation")
```
Comment (jamadden) [inline ZConfig/datatypes.py]: I'm not intimately familiar with the history there, but the linked docs still contain the phrase "When passing in a string, values for NaN and Infinity may be returned, *depending on the underlying C library*" (emphasis mine). It does say that float *represents* things a certain way, but it's not clear to me that it commits to *parsing* things a certain way.

I'd be happy to see this simplified, and that could arguably be a part of the changes for this PR. It's just not entirely clear to me its safe to do so.

File: ZConfig/info.py (hunk 1 above)
Comment (jamadden) [inline ZConfig/info.py]: I think the weirdness is partially because this is a subclass of an abstract class that allows subclasses to define the constructors differently (which they do, passing constants for some of the arguments). That comparison works on Py2, but fails on Py3.

Once we do the assert in the constructor, getdefault doesn't matter anymore?

File: ZConfig/tests/test_info.py (hunk 8 above)
Comment (mgedmin) [inline ZConfig/tests/test_info.py]: I don't recall where I learned this trick, but I recall getting burnt a few times in Zopeland when I (or somebody else) wrote a new test class and then forgot the corresponding `makeSuite()` bit.  Of course, measuring test coverage protects from that to some extent.

I also have doubts whether that refactoring belongs to this already-very-large PR (especially if we want to convert all the files in one go, rather opt for a more incremental approach and convert just those that are touched in this PR).

I have this habit where when I'm reviewing code I also tend to look for possible improvements in the surrounding codebase, and also for opportunities to share knowledge.  But then I forget to state that that's what I'm doing, rather than asking for changes in this particular PR.

File: ZConfig/components/logger/loghandler.py (hunk 11 above)
Comment (mgedmin) [inline ZConfig/components/logger/loghandler.py]: Let's just hope linters won't decide to start warning about "useless assignment X = X" in the future ;)

File: ZConfig/datatypes.py (hunk 16 above)
Comment (mgedmin) [inline ZConfig/datatypes.py]: The "depending on the underlying C library" bit remains from the earlier text, which warned that the specific inf/nan spellings did depend on the C library: http://svn.python.org/view/python/trunk/Doc/library/functions.rst?r1=59558&r2=59557&pathrev=59558.  I wonder why it wasn't removed.

The "Float accepts the strings nan, inf and -inf for NaN and positive or negative infinity.  The case and a leading + are ignored as well as a leading - is ignored for NaN." part is about the parsing.  There are [stdlib tests](https://github.com/python/cpython/blob/2.6/Lib/test/test_float.py#L268-L309) for this too.

The "Float always represents NaN and infinity as nan, inf or -inf." part is about the `repr()`.

Comment (jamadden) [inline ZConfig/datatypes.py]: Ok, cool! Thanks for the pointers. Should this change be a part of this admittedly large PR, or should it wait? ( I am willing to create follow up issues for this and other things that come out of this discussion. )

File: ZConfig/tests/test_info.py (hunk 8 above)
Comment (jamadden) [inline ZConfig/tests/test_info.py]: I was burnt a ridiculous number of times on this very PR by that issue :/ it was quickly noticed and fixed, but still. 

I'm happy to create a follow up issue. And I will be doing the same for other projects where I've dealt with this (relstorage).

File: ZConfig/datatypes.py (hunk 16 above)
Comment (mgedmin) [inline ZConfig/datatypes.py]: Separate PR, I think, to keep things simple.

File: ZConfig/info.py (hunk 1 above)
Comment (mgedmin) [inline ZConfig/info.py]: Yeah, I think we should either fully support `maxOccurs` being `None` (maybe that is how you spell "unlimited"?) by adding code to make Python 3 handle this explicitly instead of failing, or we should require it to be not-None everywhere.

I'm sorry that I don't currently have the time to investigate what `maxOccurs` is used for!

**Comment by jamadden** [comment]
I created new issues for the things that have come out in the comments. Is this mergeable?

**Comment by mgedmin** [comment]
You still have a green check from me :)

The only thing I still find kind of distasteful in the new changes is this:
- Perhaps define a helper function `raise_with_same_tb(exc)`, because I see a lot of places in the diff that do this sort of thing with new temporary exception variables (all named differently) and `reraise(type(exc), exc, sys.exc_info()[2])`.

**Comment by jamadden** [comment]
Ah, now I understand what you mean by that. That should be easy enough to change and reduce the size of the PR a bit. I'll do that.

**Comment by mgedmin** [comment]
Thank you, looks great!

File: ZConfig/tests/support.py (hunk 4 above)
Comment (freddrake) [inline ZConfig/tests/support.py]: I don't think this applies any more.

File: ZConfig/cmdline.py (hunk 10 above)
Comment (freddrake) [inline ZConfig/cmdline.py]: ``type`` has been a built-in for a very long time (since before I discovered Python at version 1.2).  I've never been particularly averse to shadowing it, though.

File: ZConfig/tests/test_datatypes.py (hunk 7 above)
Comment (freddrake) [inline ZConfig/tests/test_datatypes.py]: I can only imagine I was thinking that looking like a table would make this more readable.  I wouldn't make the same mistake now.  (I certainly hope.)

File: ZConfig/cmdline.py (hunk 10 above)
Comment (mgedmin) [inline ZConfig/cmdline.py]: FWIW when I was reviewing this, I wanted to ask why this comparison wasn't using `if bk is type` to compare with the `type` builtin.  Then I thought a bit more and decided maybe `type` was a local variable shadowing the builtin.

(Also, given that my Vim syntax-highlights `type` in a different color from regular variables makes me want to use `type_` instead of shadowing.)

File: ZConfig/components/logger/loghandler.py (hunk 11 above)
Comment (tseaver) [inline ZConfig/components/logger/loghandler.py]: - I find the reassignment useless, and would prefer a comment like `#noqa API` to make the intent clearer.
- I don't think we need to support backward-compatibility for re-exports of stdlib modules.

File: ZConfig/cmdline.py (hunk 10 above)
Comment (tseaver) [inline ZConfig/cmdline.py]: @mgedmin that was exactly what I thought when reading that diff (it took me a couple of minutes to stumble on the shadowing).
